tests/libwnbd_tests/test.cpp

```diff
@@ -463,3 +521,403 @@ TEST(TestRead, Read2PBDisk) {
+        true
+    );
+    WnbdDaemon.Start();
+
+    WNBD_CONNECTION_INFO ConnectionInfo = { 0 };
+    NTSTATUS Status = WnbdShow(InstanceName.c_str(), &ConnectionInfo);
+    ASSERT_FALSE(Status) << "couldn't retrieve WNBD disk info";
+
+    PWNBD_DISK WnbdDisk = WnbdDaemon.GetDisk();
+
+    std::string DiskPath = GetDiskPath(InstanceName);
+
+    WNBD_USR_STATS UserspaceStats = { 0 };
+    WNBD_DRV_STATS DriverStats = { 0 };
+
+    WnbdGetUserspaceStats(WnbdDisk, &UserspaceStats);
+    WnbdGetDriverStats(InstanceName.c_str(), &DriverStats);
+
+    // No requests have been sent, write counters should be 0
+    ASSERT_EQ(UserspaceStats.TotalWrittenBlocks, 0);
+    ASSERT_EQ(UserspaceStats.WriteErrors, 0);
+
+    HANDLE DiskHandle = CreateFileA(
+        DiskPath.c_str(),
+        GENERIC_WRITE | GENERIC_READ,
+        FILE_SHARE_READ | FILE_SHARE_WRITE,
+        NULL,
+        OPEN_EXISTING,
+        FILE_ATTRIBUTE_NORMAL,
+        NULL);
+    ASSERT_NE(INVALID_HANDLE_VALUE, DiskHandle)
+        << "couldn't open disk: " << DiskPath
+        << ", error: " << WinStrError(GetLastError());
+    std::unique_ptr<void, decltype(&CloseHandle)> DiskHandleCloser(
+        DiskHandle, &CloseHandle);
+
+    int WriteBufferSize = 4 << 20;
+    std::unique_ptr<void, decltype(&free)> WriteBuffer(
+        malloc(WriteBufferSize), free);
+    ASSERT_TRUE(WriteBuffer.get()) << "couldn't allocate: " << WriteBufferSize;
+    memset(WriteBuffer.get(), WRITE_BYTE_CONTENT, WriteBufferSize);
+
+    int ReadBufferSize = 4 << 20;
+    std::unique_ptr<void, decltype(&free)> ReadBuffer(
+        malloc(ReadBufferSize), free);
+    ASSERT_TRUE(ReadBuffer.get()) << "couldn't allocate: " << ReadBufferSize;
+
+    DWORD BytesRead = 0;
+
+    WNBD_IO_REQUEST ExpWnbdRequest = { 0 };
+    ExpWnbdRequest.RequestType = WnbdReqTypeWrite;
+    ExpWnbdRequest.Cmd.Write.BlockAddress = 0;
+    ExpWnbdRequest.Cmd.Write.BlockCount = 1;
+    // We expect to be the first ones writing to this disk.
+    ASSERT_FALSE(WnbdDaemon.ReqLog.HasEntry(ExpWnbdRequest));
+
+    DWORD BytesWritten = 0;
+    for (int i = 0; i < 2; i++) {
+        ASSERT_TRUE(WriteFile(
+            DiskHandle, WriteBuffer.get(),
+            DefaultBlockSize, &BytesWritten, NULL));
+        ASSERT_EQ(DefaultBlockSize, BytesWritten);
+    }
+
+    WnbdGetUserspaceStats(WnbdDisk, &UserspaceStats);
+    WnbdGetDriverStats(InstanceName.c_str(), &DriverStats);
+    ASSERT_GE(UserspaceStats.TotalWrittenBlocks, 1);
+    ASSERT_EQ(UserspaceStats.WriteErrors, 0);
+    ASSERT_GE(DriverStats.TotalReceivedIORequests, 1);
+    ASSERT_GE(DriverStats.TotalReceivedIOReplies, 1);
+
+    // Clear read buffer
+    memset(ReadBuffer.get(), 0, ReadBufferSize);
+    // 1 block
+    ASSERT_TRUE(ReadFile(
+        DiskHandle, ReadBuffer.get(),
+        DefaultBlockSize, &BytesRead, NULL)) << "Read failed: " << GetLastError() << std::endl;
+    ASSERT_EQ(DefaultBlockSize, BytesRead);
+
+    ASSERT_GE(UserspaceStats.TotalReadBlocks, 1);
```

Comment:
I'm a bit concerned that this test could be quite flaky. The IO counters get incremented *after* the reply is submitted, so we might get outdated stats. As a workaround, we're repeating the same request, but even so we'd have to exceed the number of IO workers and it can still confuse subsequent assertions.

Probably a better approach would be to have something similar to Gomega's  ``Eventually``: https://onsi.github.io/gomega/#eventually, maybe with a 1s timeout.